apply_DDT.py

```diff
@@ -0,0 +1,574 @@
+
+    # Grab the bkg data
+    bkg_cols = [Columns.load(f) for f in glob.glob(bkg-files)]
+    #bkg_cols = [Columns.load(f) for f in glob.glob(qcd_data_files)]
+    #bkg_cols.extend(Columns.load(f) for f in glob.glob(tt_data_files))
+    X, pT, mT, rho, bkg_weight = bdt_ddt_inputs(bkg_cols, lumi, all_features)
+
+    # _____________________________________________
+    # Open the trained models and get the scores
+
+    bkg_score = {}
+
+    xgb_model = xgb.XGBClassifier()
+    xgb_model.load_model(model_file)
+    with time_and_log(f'Calculating xgboost scores for {model_file}...'):
+        bkg_score = xgb_model.predict_proba(X)[:,1]
+
+    # _____________________________________________
+    # Calculate bkg efficiencies for the DDT
+
+    print("Calculating bkg efficiencies for various bdt cut values")
+    bkg_eff=[]
+    bkg_percents=[]
+    bkg_Hist={}
+    i = 0
+    bkg_Hist_nobdt=np.histogram(bkg_score[bkg_score>0.0],weights=bkg_weight[bkg_score>0.0]*len(bkg_score))
+    for bdt_cut in bdt_cuts: 
+        bkg_Hist[i]=np.histogram(bkg_score[bkg_score>bdt_cut],weights=bkg_weight[bkg_score>bdt_cut]*len(bkg_score))
+        bkg_eff.append(sum(bkg_Hist[i][0])/sum(bkg_Hist_nobdt[0]))
+        bkg_percents.append(sum(bkg_Hist[i][0])/sum(bkg_Hist_nobdt[0])*100)
+        i += 1
+
+    # _____________________________________________
+    # Create the DDT 2D map
+
+    # Only make the map if it doesn't exist
+    if not osp.exists(ddt_map_file):
+        create_DDT_map_dict(mT, pT, rho, bkg_score, bkg_weight, bkg_percents, bdt_cuts, ddt_map_file)
+    else: print("The DDT has already been calculated, please change the name if you want to remake the ddt map")
+
+    # Load the dictionary from the json file
+    with open(ddt_map_file, 'r') as f:
+        var_dict = json.load(f)
+
+    if plot_2D_DDT_map == True :
+        for key in var_dict.keys() :
+ 
+            print("Plotting the 2D DDT maps")
+ 
+            # Get the 2D DDT map and bin edges for the corresponding BDT cut (key)
+            bdt_cut = key
+            var_map_smooth, RHO_edges, PT_edges = var_dict[key]
+            var_map_smooth = np.array(var_map_smooth)
+            RHO_edges = np.array(RHO_edges)
+            PT_edges = np.array(PT_edges)
+ 
+            # Plot 2D map for rho-phi plane for each BDT cut
+            plt.figure(figsize=(10, 8))
+            hep.cms.label(rlabel="(13 TeV)")
+            plt.imshow(var_map_smooth.T, extent=[RHO_edges[0], RHO_edges[-1], PT_edges[0], PT_edges[-1]], aspect='auto', origin='lower', cmap='viridis')
+            plt.colorbar(label='DDT Map value')
+            plt.xlabel('$ \\rho = \\ln(\\mathrm{m}^{2}/\\mathrm{p}_{\\mathrm{T}}^{2})$')
+            plt.ylabel('$\\mathrm{p}_{\\mathrm{T}}$ [GeV]')
+            plt.tight_layout()
+            plt.savefig(f'plots/2D_map_cut_{bdt_cut}.png')  # Save the plot
+            plt.savefig(f'plots/2D_map_cut_{bdt_cut}.pdf')  # Save the plot
```

Comment:
define a function that takes `plt` + output name and saves both png and pdf copies, rather than repeating this pattern many times

Reply:
Done but needs some careful testing.